src/typeInference/annotateFlworExpression.ts

```diff
@@ -0,0 +1,159 @@
+ * @param ast The ast containing the letClause
+ * @param annotationContext The context in which it gets annotated and in which new variables will be introduced
+ * @param annotate the annotate function as a callback to annotate the children
+ */
+function annotateLetClause(
+	ast: IAST,
+	annotationContext: AnnotationContext,
+	annotate: (ast: IAST, annotationContext) => SequenceType
+): void {
+	const pathToVarName: string[] = ['letClauseItem', 'typedVariableBinding', 'varName'];
+	const varNameNode: IAST = astHelper.followPath(ast, pathToVarName);
+	const varName = varNameNode[1] as string;
+
+	const pathToVarType: string[] = ['letClauseItem', 'letExpr'];
+	const varTypeNode: IAST = astHelper.followPath(ast, pathToVarType);
+
+	const varType: SequenceType = annotate(varTypeNode[1] as IAST, annotationContext);
+
+	annotationContext.insertVariable(varName, varType);
+}
+
+/**
+ * A method to annotate the ForClause.
+ * @param ast the ast containing the forClause
+ * @param annotationContext the context in which its being annotated
+ * @param annotate the annotate function as a callback to annotate the children
+ */
+function annotateForClause(
+	ast: IAST,
+	annotationContext: AnnotationContext,
+	annotate: (ast: IAST, annotationContext) => SequenceType
+): void {
+	const pathToTypedVariableBinding = ['forClauseItem', 'typedVariableBinding', 'varName'];
+	const pathToForExpr = ['forClauseItem', 'forExpr', 'sequenceExpr'];
+	// const pathToForBody = ['forClauseItem', 'for'];
+	const varName = astHelper.followPath(ast, pathToTypedVariableBinding)[1] as string;
+
+	const varTypeNode: IAST = astHelper.followPath(ast, pathToForExpr);
+	if (!varTypeNode) {
+		return;
+	}
+
+	// A set of all the SequenceType in the sequenceExpr
+	const allTypes: SequenceType[] = astHelper
+		.getChildren(varTypeNode, '*')
+		.map((element) => annotate(element, annotationContext));
+
+	// filter on unique types
+	const types = filterOnUniqueObjects(allTypes);
+
+	// if there is only 1 unique type we can add the variable to have this type.
+	if (types.length === 1) {
+		annotationContext.insertVariable(varName, types[0]);
+	}
+}
+
+/**
+ * Annotate the where clause.
+ * @param ast the ast containing the whereClause
+ * @param annotationContext the context in which its being annotated
+ * @param annotate the annotate function as a callback to annotate the children
+ * @returns always returns exactly 1 boolean
+ */
+function annotateWhereClause(
+	ast: IAST,
+	annotationContext: AnnotationContext,
+	annotate: (ast: IAST, annotationContext) => SequenceType
+): SequenceType {
+	const seqType = {
+		type: ValueType.XSBOOLEAN,
+		mult: SequenceMultiplicity.EXACTLY_ONE,
+	};
+
+	// Annotate the child nodes
+	annotate(ast, annotationContext);
+
+	astHelper.insertAttribute(ast, 'type', seqType);
+	return seqType;
+}
+
+function annotateOrderByClause(ast: IAST, annotationContext: AnnotationContext): undefined {
+	return undefined;
```

Comment:
Does this one not get a type?

Reply:
The order by does not get a type no, you prefer if we just leave this out?